### PR TITLE
Streaming Encoder/Decoder

### LINKED ISSUE #5: feature request: asynchronous API for decoding
Hi, I'd like to use COBS in the [framed-serial](https://github.com/astraw/framed-serial) crate, but to do so, I need asynchronous decoding. Specifically, I want to pass in each incoming byte as it arrives to check for the completion of a valid frame.

### PR DESCRIPTION
**Pull request by jamesmunns**
Closes #5. Should work even for no_std targets!

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,5 +1,70 @@
...
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Complete encoding of the output message. Does NOT terminate
+    /// the message with the sentinel value
+    pub fn finalize(self) -> Result<usize, ()> {
```
Comment (jamesmunns) [inline src/lib.rs]: @awelkie do you think that the `finalize()` method should insert the final `0x00`? I don't think I did this, as it didn't match your current API, but IMO it makes sense to do.

File: src/lib.rs [hunk 2]
```diff
@@ -76,6 +114,176 @@ pub fn encode_vec_with_sentinel(source: &[u8], sentinel: u8) -> Vec<u8> {
...
+                self.dest_idx += 1;
+                (Ok(None), Grab(i - 1))
+            },
+
+            // We have reached the end of a data run, and we expect a data header
+            // here. Since 0x00 is not a valid overhead byte, this is an error case.
+            // This may non *NEED* to be an error, but Example 9 on
+            // https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing
+            // indicates that this is not correct behavior
+            (GrabChain(0), 0x00) => {
```
Comment (jamesmunns) [inline src/lib.rs]: Hey @awelkie, let me know what you think about this state transition. Example 9 of https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing shows that they expect an extra overhead byte after a 254 byte "GrabChain", even if it signifies that the next byte is a zero.

I think this is superfluous, but maybe I am missing something.

Once we have a decision, I can write a test that works one way and fails the other, so the ambiguity is resolved.

Comment (jamesmunns) [reply src/lib.rs]: Oops, I misunderstood the test set. There is a trailing zero that causes this to happen. I'll correct the code.

**Comment by jamesmunns** [comment]
Okay, I'm done with messing with this PR now, let me know if you have any questions!

File: src/lib.rs (hunk 1 above)
Comment (awelkie) [reply src/lib.rs]: Yeah good question. I didn't insert `0x00` at the beginning or the end to let the user decide where to insert the sentinel. I figure there may be systems where the user would want to insert the sentinel at the beginning of the frame and not the end, or vice-versa, or maybe both, so I just avoided making that decision.

I'd prefer to keep it that way, but I also realize that in nearly every situation one would want to insert the sentinel at the end, so I'm open to debate.

**Comment by awelkie** [comment]
LGTM. Ideally I think the streaming decoder would allow streaming the output as well, instead of placing the output in a buffer. So with every input byte the decoder would return zero or one output bytes, and the user could consume the output in a streaming manner. But this PR works and is an improvement to the API so I'll merge it.

**Pull request merged**